test/Learners/LearnersTests/MatchboxRecommenderTests.cs

```diff
@@ -726,53 +726,47 @@ string SerializeUserOrItem(object obj)
-                {
-                    using (var writer = new DelegatingWrappedBinaryWriter(new BinaryWriter(stream, Encoding.UTF8, true), SerializeUserOrItem))
-                    {
-                        recommender.SaveForwardCompatible(writer);
-                    }
-                }
-                recommender.Train(this.standardTrainingData, this.standardTrainingDataFeatures);
 
-                // Cannot just call recommender.SaveForwardCompatible(TrainedFileName); because we need to specially handle Users and Items.
-                using (Stream stream = File.Open(TrainedFileName, FileMode.Create))
-                {
-                    using (var writer = new DelegatingWrappedBinaryWriter(new BinaryWriter(stream, Encoding.UTF8, true), SerializeUserOrItem))
-                    {
-                        recommender.SaveForwardCompatible(writer);
-                    }
-                }
+                SaveRecommender(NotTrainedFileName, recommender);
+                recommender.Train(this.standardTrainingData, this.standardTrainingDataFeatures);
+                SaveRecommender(TrainedFileName, recommender);
 
                 CheckStandardRatingPrediction(recommender, this.standardTrainingDataFeatures);
             }
 
             // Deserialize and test
             {
-                // Cannot just call MatchboxRecommender.LoadBackwardCompatible<StandardDataset, Tuple< User, Item, int?>, User, Item, int, FeatureProvider>(TrainedFileName, this.standardMapping); because we need to specially handle Users and Items.
-                IMatchboxRecommender<StandardDataset, User, Item, RatingDistribution, FeatureProvider> trainedRecommender;
-                using (Stream stream = File.Open(TrainedFileName, FileMode.Open))
+                var trainedRecommender = LoadRecommender(TrainedFileName);
+                var notTrainedRecommender = LoadRecommender(NotTrainedFileName);
+
+                notTrainedRecommender.Train(this.standardTrainingData, this.standardTrainingDataFeatures);
+
+                CheckStandardRatingPrediction(trainedRecommender, this.standardTrainingDataFeatures);
+                CheckStandardRatingPrediction(notTrainedRecommender, this.standardTrainingDataFeatures);
+            }
+
+            // Cannot just call recommender.SaveForwardCompatible(TrainedFileName); because we need to specially handle Users and Items.
+            void SaveRecommender(string fileName, IMatchboxRecommender<StandardDataset, User, Item, RatingDistribution, FeatureProvider> recommender)
+            {
+                using (Stream stream = File.Open(fileName, FileMode.Create))
                 {
-                    using (var reader = new DelegatingWrappedBinaryReader(new BinaryReader(stream, Encoding.UTF8, true), new DeserializeUserOrItem()))
+                    using (var writer = new DelegatingWrappedBinaryWriter(new BinaryWriter(stream, Encoding.UTF8, true), SerializeUserOrItem))
```

Comment:
Are users expected to do this too?  If so, why not provide them with the tools to do it?

Reply:
They can do exactly the same thing I did here so what sort of additional tools do you mean?

Do you mean moving these loading and saving functions to the library and making them generic? It would be possible to move this loading and saving function into the runtime.

Either way the user would still need to implement an interface if we want to keep the nice open-generic method. Otherwise could switch to a delegate like Func<string, Type, object> (where we tell the implementation what type is expected -- maybe that would be better anyway).

Reply:
How can they do 'exactly the same thing' when DelegatingWrappedBinaryReader is internal?

Reply:
They can provide their own.

Are you saying you'd like DelegatingWrappedBinaryReader to be public (so they can be used by others), or that you want these Save/Load functions to be public and generic (so they can be used by others), or both?

Reply:
Either is fine with me.

Reply:
Added